client/light/src/fetcher.rs

```diff
@@ -201,6 +204,35 @@ impl<E, H, B: BlockT, S: BlockchainStorage<B>> LightDataChecker<E, H, B, S> {
 			Ok(storage)
 		}, storage)
 	}
+
+	fn code_and_heap_pages_from_cache(&self, block: &BlockId<B>) -> ClientResult<(Vec<u8>, u64)> {
+		let cache = self.blockchain.cache().unwrap();
+		let result = cache.get_at(&well_known_cache_keys::HEAP_PAGES, block)
+			.and_then(|heap_pages| {
+				let code_hash = cache.get_at(&well_known_cache_keys::CODE_HASH, block)?;
+				match (code_hash, heap_pages) {
+					(Some((_, _, code_hash)), Some((_, _, heap_pages))) => {
+						let heap_pages: u64 = Decode::decode(&mut &heap_pages[..])
+							.expect("heap pages is properly encoded when stored in cache; qed");
+						Ok((code_hash, heap_pages))
+					}
+					_ => Err(sp_blockchain::Error::Msg("couldn't fetch code_hash and heap_pages".into()))
+				}
+			})
+			.and_then(|(code_hash, heap_pages)| {
+				let mut key = Vec::new();
+				key.extend_from_slice(b"runtime-code");
+				key.extend_from_slice(&code_hash);
+				if let Some(runtime_code) = self.blockchain.storage().get_aux(&key)? {
+					return Ok(Some((runtime_code, heap_pages)))
+				}
+				Ok(None)
+			})?
+			.ok_or_else(|| {
+				sp_blockchain::Error::Msg(format!("couldn't fetch code_hash and heap_pages"))
+			})?;
+		Ok(result)
```

Comment:
I don't like this `and_then` nesting. It would be much better to just have something like:
```rust
let error = || {
    sp_blockchain::Error::Msg(format!("couldn't fetch code_hash and heap_pages"))
};

let heap_pages = cache.get_at(&well_known_cache_keys::HEAP_PAGES, block).ok_or_else(error)?;
let code_hash = cache.get_at(&well_known_cache_keys::CODE_HASH, block).ok_or_else(error)?;
```